pylops/signalprocessing/nonstatconvolve2d.py

```diff
@@ -292,3 +295,347 @@ def _rmatvec(self, x: NDArray) -> NDArray:
+        not (``False``)
+
+    Raises
+    ------
+    ValueError
+        If filters ``hs`` have even size
+    ValueError
+        If ``ihx`` or ``ihz`` is not regularly sampled
+    NotImplementedError
+        If ``engine`` is neither ``numpy``, ``fftw``, nor ``scipy``.
+
+    Notes
+    -----
+    The NonStationaryConvolve2D operator is used to estimate a non-stationary
+    two-dimensional filter betwen two signals, an input signal (provided directly
+    to the operator) and the desired output signal.
+
+    For more details on the numerical implementation of the forward and adjoint,
+    see :class:`pylops.signalprocessing.NonStationaryFilters1D`.
+
+    """
+
+    def __init__(
+        self,
+        inp: NDArray,
+        hshape: InputDimsLike,
+        ihx: InputDimsLike,
+        ihz: InputDimsLike,
+        engine: str = "numpy",
+        num_threads_per_blocks: Tuple[int, int] = (32, 32),
+        dtype: DTypeLike = "float64",
+        name: str = "C",
+    ) -> None:
+        if engine not in ["numpy", "numba", "cuda"]:
+            raise NotImplementedError("engine must be numpy or numba or cuda")
+        if hshape[0] % 2 == 0 or hshape[1] % 2 == 0:
+            raise ValueError("filters hs must have odd length")
+        if len(np.unique(np.diff(ihx))) > 1 or len(np.unique(np.diff(ihz))) > 1:
+            raise ValueError(
+                "the indices of filters 'ih' are must be regularly sampled"
+            )
+
+        self.inp = inp
+        self.inpdims = inp.shape
+        self.hshape = hshape
+        self.ohx, self.dhx, self.nhx = ihx[0], ihx[1] - ihx[0], len(ihx)
+        self.ohz, self.dhz, self.nhz = ihz[0], ihz[1] - ihz[0], len(ihz)
+        self.ehx, self.ehz = ihx[-1], ihz[-1]
+        self.engine = engine
+        super().__init__(
+            dtype=np.dtype(dtype),
+            dims=(self.nhx, self.nhz, *hshape),
+            dimsd=self.inpdims,
+            name=name,
+        )
+
+        # create additional input parameters for engine=cuda
+        self.kwargs_cuda = {}
+        if engine == "cuda":
+            self.kwargs_cuda["num_threads_per_blocks"] = num_threads_per_blocks
+            num_threads_per_blocks_x, num_threads_per_blocks_z = num_threads_per_blocks
+            num_blocks_x = (
+                self.dims[0] + num_threads_per_blocks_x - 1
+            ) // num_threads_per_blocks_x
+            num_blocks_z = (
+                self.dims[1] + num_threads_per_blocks_z - 1
+            ) // num_threads_per_blocks_z
+            self.kwargs_cuda["num_blocks"] = (num_blocks_x, num_blocks_z)
+        self._register_multiplications(engine)
+
+    def _register_multiplications(self, engine: str) -> None:
+        if engine == "numba":
+            numba_opts = dict(nopython=True, fastmath=True, nogil=True, parallel=True)
+            self._mv = jit(**numba_opts)(self.__matvec)
+            # Currently a race condition seem to occur when updating parts of hs multiple times within
```

Comment:
@cako if you have time could have take a look into this… initially I though I wrote the adjoint wrong, but once I switched to pure python everything worked fine. I am not sure I understand how this differ from a similar code pattern in the NonStatConvolve2D operator (the only maybe important difference is that here I need to do 4 updates of h instead of just one of x

Reply:
Right, so it appears you have a line like this: `hs[f(ix), g(iz)] += something`. This will likely create a race condition which you noticed. One solution would be an atomic CPU op (which [doesn't exist so far in Numba](https://github.com/numba/numba/issues/2988)). Otherwise you can create separate vectors for each ix, and then stack them all non-parallel at the end of the parallel for. So you'd have something like this:
```python
tmphs = zeros((nx, *hs.shape))
for ix in prange:
    for iz in range:
        tmphs[ix, f(ix), g(iz)] += something
return tmphs.sum(axis=0)
```

Reply:
Mmh right, I see your point and I think this agrees with what I have read in the numba link I put in the comment about race condition when updating slices of the vector.

However I am then confused why this one works then, https://github.com/PyLops/pylops/blob/92191c82fb56b55e9fba5425fd003e941616e175/pylops/signalprocessing/nonstatconvolve2d.py#L248 ???

I can implement your suggestion, for 3d I am worried it may be very memory consuming but for 2d I think it’s doable